File: app/routers/purchases.py
from datetime import datetime

from fastapi import APIRouter, HTTPException, Depends
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.models.product import Product
from app.models.purchase import Purchase
from app.models.purchase_item import PurchaseItem
from app.models.supplier import Supplier
from app.models.supplier_payment import SupplierPayment
from app.models.shop_operation import ShopOperation
from app.rbac import require_permission
from app.schemas.purchase import PurchaseCreate, PurchaseRead, CancelPurchasePayload
from app.services.shop_context_service import adjust_stock, get_current_shop_id, get_effective_stock, record_shop_operation

router = APIRouter(tags=["achats"])


def _purchases_query(db: Session):
    query = db.query(Purchase)
    shop_id = get_current_shop_id(db)
    if shop_id is None:
        return query
    return query.join(
        ShopOperation,
        (ShopOperation.entity_type == "purchase")
        & (ShopOperation.entity_id == Purchase.id),
    ).filter(ShopOperation.shop_id == shop_id)


def _purchase_in_current_shop(db: Session, purchase_id: int):
    return _purchases_query(db).filter(Purchase.id == purchase_id).first()


def add_event(
    db: Session,
    entity_type: str,
    entity_id: int,
    event_type: str,
    amount_signed: int | None = None,
    note: str | None = None,
):
    from app.models.transaction_event import TransactionEvent

    db.add(
        TransactionEvent(
            entity_type=entity_type,
            entity_id=entity_id,
            event_type=event_type,
            amount_signed=amount_signed,
            note=note,
        )
    )


def add_stock_movement(
    db: Session,
    product_id: int,
    movement_type: str,
    quantity: int,
    reference_type: str,
    reference_id: int,
    note: str | None = None,
):
    from app.models.stock_movement import StockMovement

    db.add(
        StockMovement(
            product_id=product_id,
            movement_type=movement_type,
            quantity=quantity,
            reference_type=reference_type,
            reference_id=reference_id,
            note=note,
        )
    )


@router.get("/purchases", response_model=list[PurchaseRead])
def list_purchases(
    db: Session = Depends(get_db),
    _allowed: None = Depends(require_permission("purchase.read")),
):
    """Liste les achats de la boutique active, du plus récent au plus ancien."""
    return _purchases_query(db).order_by(Purchase.created_at.desc(), Purchase.id.desc()).all()


@router.post("/purchases", response_model=PurchaseRead)
def create_purchase(
    payload: PurchaseCreate,
    db: Session = Depends(get_db),
    _allowed: None = Depends(require_permission("purchase.create")),
):
    """Crée un achat multi-produits, augmente le stock et met à jour la dette fournisseur."""
    supplier = db.query(Supplier).filter(Supplier.id == payload.supplier_id).first()
    if not supplier:
        raise HTTPException(status_code=404, detail="Fournisseur introuvable")

    if not payload.items:
        raise HTTPException(status_code=400, detail="Au moins une ligne produit est requise")

    total_amount = 0
    resolved_items = []

    for item in payload.items:
        if item.quantity <= 0:
            raise HTTPException(status_code=400, detail="La quantité doit être supérieure à zéro")

        if item.unit_cost < 0:
            raise HTTPException(status_code=400, detail="Le coût unitaire ne peut pas être négatif")

        product = db.query(Product).filter(Product.id == item.product_id).first()
        if not product:
            raise HTTPException(status_code=404, detail=f"Produit introuvable : {item.product_id}")

        line_total = item.unit_cost * item.quantity
        total_amount += line_total
        resolved_items.append((product, item.quantity, item.unit_cost, line_total))

    paid_amount = payload.paid_amount
    if paid_amount < 0:
        raise HTTPException(status_code=400, detail="Le montant payé ne peut pas être négatif")

    if paid_amount > total_amount:
        raise HTTPException(
            status_code=400,
            detail="Le montant payé ne peut pas dépasser le montant total",
        )

    remaining_amount = total_amount - paid_amount

    if remaining_amount > 0 and payload.due_date is None:
        raise HTTPException(
            status_code=400,
            detail="Une date d'échéance est obligatoire pour une dette fournisseur",
        )

    if remaining_amount == 0:
        status = "paid"
    elif paid_amount == 0:
        status = "credit"
    else:
        status = "partial"

    purchase = Purchase(
        supplier_id=payload.supplier_id,
        total_amount=total_amount,
        paid_amount=paid_amount,
        remaining_amount=remaining_amount,
        status=status,
        due_date=payload.due_date,
        original_amount=(
            payload.original_amount
            if payload.original_amount is not None
            else total_amount
        ),
        original_currency=(
            payload.original_currency or "XOF"
        ).upper(),
        exchange_rate=payload.exchange_rate,
    )
    db.add(purchase)
    db.flush()
    record_shop_operation("purchase", purchase.id, db)

    for product, quantity, unit_cost, line_total in resolved_items:
        adjust_stock(product, quantity, db)

        db.add(
            PurchaseItem(
                purchase_id=purchase.id,
                product_id=product.id,
                quantity=quantity,
                unit_cost=unit_cost,
                line_total=line_total,
                paid_amount=0,
                remaining_amount=line_total,
                status="credit",
            )
        )

        add_stock_movement(
            db=db,
            product_id=product.id,
            movement_type="purchase_in",
            quantity=quantity,
            reference_type="purchase",
            reference_id=purchase.id,
            note=f"Achat chez le fournisseur {supplier.name}",
        )

    if paid_amount > 0:
        db.add(
            SupplierPayment(
                purchase_id=purchase.id,
                supplier_id=supplier.id,
                amount=paid_amount,
                channel=payload.payment_channel,
                reference=None,
            )
        )

    supplier.debt += remaining_amount

    add_event(
        db,
        "purchase",
        purchase.id,
        "created",
        amount_signed=total_amount,
        note="Achat créé",
    )

    db.commit()
    db.refresh(purchase)
    return purchase


@router.get("/purchases/{purchase_id}/items")
def get_purchase_items(
    purchase_id: int,
    db: Session = Depends(get_db),
    _allowed: None = Depends(require_permission("purchase.read")),
):
    """Affiche les lignes produit d’un achat de la boutique active."""
    purchase = _purchase_in_current_shop(db, purchase_id)
    if not purchase:
        raise HTTPException(status_code=404, detail="Achat introuvable")

    return db.query(PurchaseItem).filter(PurchaseItem.purchase_id == purchase_id).all()


@router.get("/purchases/{purchase_id}/payments")
def get_purchase_payments(
    purchase_id: int,
    db: Session = Depends(get_db),
    _allowed: None = Depends(require_permission("purchase.read")),
):
    """Affiche les paiements liés à un achat de la boutique active."""
    purchase = _purchase_in_current_shop(db, purchase_id)
    if not purchase:
        raise HTTPException(status_code=404, detail="Achat introuvable")

    return db.query(SupplierPayment).filter(SupplierPayment.purchase_id == purchase_id).all()


@router.post("/purchases/{purchase_id}/cancel", response_model=PurchaseRead)
def cancel_purchase(
    purchase_id: int,
    payload: CancelPurchasePayload,
    db: Session = Depends(get_db),
    _allowed: None = Depends(require_permission("purchase.cancel")),
):
    """Annule un achat sans le supprimer, corrige le stock et la dette fournisseur."""
    purchase = _purchase_in_current_shop(db, purchase_id)
    if not purchase:
        raise HTTPException(status_code=404, detail="Achat introuvable")

    if purchase.status == "cancelled":
        raise HTTPException(status_code=400, detail="Cet achat est déjà annulé")

    purchase_items = db.query(PurchaseItem).filter(PurchaseItem.purchase_id == purchase.id).all()
    if not purchase_items:
        raise HTTPException(status_code=400, detail="Cet achat ne contient aucune ligne")

    supplier = None
    if purchase.supplier_id:
        supplier = db.query(Supplier).filter(Supplier.id == purchase.supplier_id).first()

    for item in purchase_items:
        product = db.query(Product).filter(Product.id == item.product_id).first()
        if not product:
            continue

        if get_effective_stock(product, db) < item.quantity:
            raise HTTPException(
                status_code=400,
                detail=f"Impossible d’annuler l’achat : stock insuffisant pour le produit {product.name}",
            )

        adjust_stock(product, -item.quantity, db)

        add_stock_movement(
            db=db,
            product_id=product.id,
            movement_type="purchase_cancel_reversal",
            quantity=-item.quantity,
            reference_type="purchase",
            reference_id=purchase.id,
            note=payload.reason,
        )

    if supplier:
        supplier.debt = max(0, supplier.debt - purchase.remaining_amount)

    purchase.status = "cancelled"
    purchase.remaining_amount = 0
    purchase.cancelled_at = datetime.utcnow()
    purchase.cancellation_reason = payload.reason

    add_event(
        db,
        "purchase",
        purchase.id,
        "cancelled",
        amount_signed=-purchase.total_amount,
        note=payload.reason,
    )

    db.commit()
    db.refresh(purchase)
    return purchase

File: app/pwa/purchases.js
(() => {
  const purchaseCart = new Map();
  let suppliers = [];
  let purchases = [];
  let selectedPayablePurchaseId = null;
  let loadedShopId = null;
  let loadingPurchasingData = false;

  const money = (value) => fmt(Number(value || 0));

  function localDateKey(date = new Date()) {
    const yyyy = date.getFullYear();
    const mm = String(date.getMonth() + 1).padStart(2, '0');
    const dd = String(date.getDate()).padStart(2, '0');
    return yyyy + '-' + mm + '-' + dd;
  }

  function setPurchaseDueDays(days) {
    const input = $('purchaseDueDate');
    if (!input) return;
    const date = new Date();
    date.setDate(date.getDate() + Number(days || 0));
    input.value = localDateKey(date);
  }

  function purchaseTotal() {
    return [...purchaseCart.values()].reduce(
      (sum, line) => sum + Number(line.quantity || 0) * Number(line.unit_cost || 0),
      0,
    );
  }

  function syncPurchaseSelectors() {
    const supplierSelect = $('purchaseSupplier');
    const productSelect = $('purchaseProduct');

    if (supplierSelect) {
      const selected = supplierSelect.value;
      supplierSelect.innerHTML =
        '<option value="">Choisir un fournisseur</option>' +
        suppliers
          .map((supplier) =>
            '<option value="' + supplier.id + '">' + esc(supplier.name) + '</option>'
          )
          .join('');
      if ([...supplierSelect.options].some((option) => option.value === selected)) {
        supplierSelect.value = selected;
      }
    }

    if (productSelect) {
      const selected = productSelect.value;
      productSelect.innerHTML =
        '<option value="">Choisir un produit</option>' +
        state.products
          .map((product) =>
            '<option value="' + product.id + '">' +
            esc(product.name) + ' · stock ' + Number(product.stock || 0) +
            '</option>'
          )
          .join('');
      if ([...productSelect.options].some((option) => option.value === selected)) {
        productSelect.value = selected;
      }
    }
  }

  function renderPurchaseCart() {
    const box = $('purchaseCart');
    if (!box) return;

    const lines = [...purchaseCart.values()];
    box.innerHTML = lines.length
      ? lines.map((line) => {
          const lineTotal = Number(line.quantity) * Number(line.unit_cost);
          return `
            <article class="purchase-line">
              <div>
                <strong>${esc(line.product.name)}</strong>
                <span>${line.quantity} × ${money(line.unit_cost)}</span>
              </div>
              <div class="purchase-line-end">
                <strong>${money(lineTotal)}</strong>
                <button type="button" class="ghost" data-purchase-remove="${line.product.id}">Retirer</button>
              </div>
            </article>
          `;
        }).join('')
      : '<p class="muted empty-state">Aucune ligne d’achat.</p>';

    $('purchaseTotal').textContent = money(purchaseTotal());

    const total = purchaseTotal();
    const paidInput = $('purchasePaidAmount');
    if (paidInput) {
      paidInput.max = String(total);
      if (Number(paidInput.value || 0) > total) {
        paidInput.value = String(total);
      }
    }
  }

  function supplierName(id) {
    return suppliers.find((supplier) => Number(supplier.id) === Number(id))?.name || 'Fournisseur';
  }

  function purchaseStatusLabel(status) {
    const labels = {
      paid: 'Payé',
      credit: 'À crédit',
      partial: 'Partiel',
      cancelled: 'Annulé',
    };
    return labels[status] || status || '';
  }

  function renderSupplierWorkspace() {
    if ($('supplierCount')) $('supplierCount').textContent = String(suppliers.length);
    if ($('supplierDebtTotal')) {
      $('supplierDebtTotal').textContent = money(
        suppliers.reduce((sum, supplier) => sum + Number(supplier.debt || 0), 0)
      );
    }

    if ($('supplierCreateCard')) {
      $('supplierCreateCard').hidden = !can('supplier.create');
    }

    if ($('supplierList')) {
      $('supplierList').innerHTML = suppliers.length
        ? suppliers.map((supplier) => `
            <article class="item">
              <div class="item-main">
                <div class="item-title">${esc(supplier.name)}</div>
                <div class="item-meta">${esc(supplier.phone || 'Sans téléphone')}</div>
              </div>
              <div class="money">Dette ${money(supplier.debt)}</div>
            </article>
          `).join('')
        : '<p class="muted empty-state">Aucun fournisseur.</p>';
    }
  }

  function renderPurchases() {
    const today = localDateKey();

    if ($('purchaseList')) {
      $('purchaseList').innerHTML = purchases.length
        ? purchases.slice(0, 20).map((purchase) => `
            <article class="purchase-history-row">
              <div>
                <strong>${esc(supplierName(purchase.supplier_id))}</strong>
                <span>Achat #${purchase.id} · ${esc(purchaseStatusLabel(purchase.status))}</span>
                <small>${purchase.created_at ? esc(fmtSaleDate(purchase.created_at)) : ''}</small>
              </div>
              <div class="purchase-line-end">
                <strong>${money(purchase.total_amount)}</strong>
                <small>Reste ${money(purchase.remaining_amount)}</small>
              </div>
            </article>
          `).join('')
        : '<p class="muted empty-state">Aucun achat.</p>';
    }

    const payables = purchases
      .filter((purchase) =>
        Number(purchase.remaining_amount || 0) > 0 &&
        String(purchase.status) !== 'cancelled'
      )
      .sort((a, b) => {
        const dueA = a.due_date || '9999-12-31';
        const dueB = b.due_date || '9999-12-31';
        if (dueA !== dueB) return dueA.localeCompare(dueB);
        return Number(b.id) - Number(a.id);
      });

    if ($('supplierPayablesList')) {
      $('supplierPayablesList').innerHTML = payables.length
        ? payables.map((purchase) => {
            const overdue = purchase.due_date && purchase.due_date < today;
            const dueToday = purchase.due_date === today;
            const badge = overdue
              ? '<span class="receivable-state receivable-overdue">En retard</span>'
              : dueToday
                ? '<span class="receivable-state receivable-today">Aujourd’hui</span>'
                : '<span class="receivable-state receivable-upcoming">À venir</span>';

            return `
              <article class="receivable-row">
                <div class="receivable-main">
                  <div class="receivable-title">
                    <strong>${esc(supplierName(purchase.supplier_id))}</strong>
                    ${badge}
                  </div>
                  <span>Achat #${purchase.id} · échéance ${esc(fmtDueDate(purchase.due_date))}</span>
                  <small>Total ${money(purchase.total_amount)} · payé ${money(purchase.paid_amount)}</small>
                </div>
                <div class="receivable-end">
                  <strong>${money(purchase.remaining_amount)}</strong>
                  ${can('supplier_payment.create')
                    ? '<button type="button" class="primary receivable-pay-btn" data-supplier-pay="' + purchase.id + '">Payer</button>'
                    : ''}
                </div>
              </article>
            `;
          }).join('')
        : '<p class="muted empty-state">Aucune dette fournisseur ouverte.</p>';
    }
  }

  async function loadPurchasingData() {
    if (!token || loadingPurchasingData) return;
    loadingPurchasingData = true;
    try {
      const results = await Promise.all([
        can('supplier.read') ? api('/pwa/suppliers') : Promise.resolve([]),
        can('purchase.read') ? api('/pwa/purchases') : Promise.resolve([]),
      ]);
      suppliers = results[0] || [];
      purchases = results[1] || [];
      loadedShopId = state.merchant?.shop_id ?? null;
      syncPurchaseSelectors();
      renderSupplierWorkspace();
      renderPurchases();
      renderPurchaseCart();
    } catch (error) {
      toast(error.message || 'Impossible de charger les achats.');
    } finally {
      loadingPurchasingData = false;
    }
  }

  $('supplierForm')?.addEventListener('submit', async (event) => {
    event.preventDefault();
    if (!can('supplier.create')) return toast('Permission insuffisante.');

    try {
      await api('/pwa/suppliers', {
        method: 'POST',
        body: JSON.stringify({
          name: $('supplierName').value.trim(),
          phone: $('supplierPhone').value.trim() || null,
          debt: 0,
        }),
      });
      event.target.reset();
      await loadPurchasingData();
      toast('Fournisseur ajouté.');
    } catch (error) {
      toast(error.message);
    }
  });

  $('purchaseAddLineBtn')?.addEventListener('click', () => {
    const productId = Number($('purchaseProduct').value);
    const product = state.products.find((item) => Number(item.id) === productId);
    const quantity = Number($('purchaseQty').value);
    const unitCost = Number($('purchaseUnitCost').value);

    if (!product) return toast('Choisis un produit.');
    if (!Number.isFinite(quantity) || quantity <= 0) return toast('Quantité invalide.');
    if (!Number.isFinite(unitCost) || unitCost < 0) return toast('Coût unitaire invalide.');

    purchaseCart.set(product.id, {
      product,
      quantity,
      unit_cost: unitCost,
    });

    $('purchaseQty').value = '1';
    $('purchaseUnitCost').value = String(product.purchase_price || 0);
    renderPurchaseCart();
  });

  $('purchaseProduct')?.addEventListener('change', () => {
    const product = state.products.find(
      (item) => Number(item.id) === Number($('purchaseProduct').value)
    );
    if (product) $('purchaseUnitCost').value = String(product.purchase_price || 0);
  });

  document.addEventListener('click', (event) => {
    const remove = event.target.closest('[data-purchase-remove]');
    if (remove) {
      purchaseCart.delete(Number(remove.dataset.purchaseRemove));
      renderPurchaseCart();
      return;
    }

    const due = event.target.closest('[data-purchase-due-days]');
    if (due) {
      setPurchaseDueDays(due.dataset.purchaseDueDays);
      return;
    }

    const pay = event.target.closest('[data-supplier-pay]');
    if (pay) {
      openSupplierPayment(Number(pay.dataset.supplierPay));
    }
  });

  $('purchaseSubmitBtn')?.addEventListener('click', async () => {
    if (!can('purchase.create')) return toast('Permission insuffisante.');
    if (!purchaseCart.size) return toast('Ajoute au moins un produit.');

    const supplierId = Number($('purchaseSupplier').value);
    if (!supplierId) return toast('Choisis un fournisseur.');

    const total = purchaseTotal();
    const paidAmount = Number($('purchasePaidAmount').value || 0);
    if (!Number.isFinite(paidAmount) || paidAmount < 0 || paidAmount > total) {
      return toast('Montant payé invalide.');
    }

    const remaining = total - paidAmount;
    const dueDate = $('purchaseDueDate').value || null;
    if (remaining > 0 && !dueDate) {
      return toast('Choisis une échéance pour la dette fournisseur.');
    }

    const button = $('purchaseSubmitBtn');
    button.disabled = true;

    try {
      await api('/pwa/purchases', {
        method: 'POST',
        body: JSON.stringify({
          supplier_id: supplierId,
          items: [...purchaseCart.values()].map((line) => ({
            product_id: line.product.id,
            quantity: line.quantity,
            unit_cost: line.unit_cost,
          })),
          paid_amount: paidAmount,
          payment_channel: $('purchasePaymentChannel').value,
          due_date: remaining > 0 ? dueDate : null,
          original_amount: total,
          original_currency: state.currencyContext?.shop_currency || 'XOF',
          exchange_rate: null,
        }),
      });

      purchaseCart.clear();
      $('purchasePaidAmount').value = '0';
      $('purchaseDueDate').value = '';
      await refresh();
      await loadPurchasingData();
      toast('Achat enregistré. Stock mis à jour.');
    } catch (error) {
      toast(error.message || 'Impossible d’enregistrer l’achat.');
    } finally {
      button.disabled = false;
    }
  });

  function openSupplierPayment(purchaseId) {
    const purchase = purchases.find((item) => Number(item.id) === Number(purchaseId));
    if (!purchase) return toast('Achat introuvable.');

    selectedPayablePurchaseId = purchase.id;
    $('supplierPaymentTitle').textContent = 'Paiement · ' + supplierName(purchase.supplier_id);
    $('supplierPaymentMeta').textContent =
      'Achat #' + purchase.id + ' · reste ' + money(purchase.remaining_amount) +
      ' · échéance ' + fmtDueDate(purchase.due_date);
    $('supplierPaymentAmount').value = String(purchase.remaining_amount);
    $('supplierPaymentAmount').max = String(purchase.remaining_amount);
    $('supplierPaymentReference').value = '';
    $('supplierPaymentSheet').hidden = false;
  }

  function closeSupplierPayment() {
    $('supplierPaymentSheet').hidden = true;
    selectedPayablePurchaseId = null;
  }

  $('supplierPaymentCloseBtn')?.addEventListener('click', closeSupplierPayment);

  $('supplierPaymentSubmitBtn')?.addEventListener('click', async () => {
    const purchase = purchases.find(
      (item) => Number(item.id) === Number(selectedPayablePurchaseId)
    );
    if (!purchase) return toast('Achat introuvable.');

    const amount = Number($('supplierPaymentAmount').value);
    if (!Number.isFinite(amount) || amount <= 0) return toast('Montant invalide.');
    if (amount > Number(purchase.remaining_amount || 0)) {
      return toast('Le paiement dépasse le reste dû.');
    }

    const button = $('supplierPaymentSubmitBtn');
    button.disabled = true;
    try {
      await api('/pwa/supplier-payments', {
        method: 'POST',
        body: JSON.stringify({
          purchase_id: purchase.id,
          supplier_id: purchase.supplier_id,
          amount,
          channel: $('supplierPaymentChannel').value,
          reference: $('supplierPaymentReference').value.trim() || null,
        }),
      });

      closeSupplierPayment();
      await loadPurchasingData();
      toast(
        amount === Number(purchase.remaining_amount || 0)
          ? 'Dette fournisseur soldée.'
          : 'Paiement fournisseur enregistré.'
      );
    } catch (error) {
      toast(error.message || 'Impossible d’enregistrer le paiement.');
    } finally {
      button.disabled = false;
    }
  });

  const originalRender = render;
  render = function() {
    originalRender();
    syncPurchaseSelectors();
    renderSupplierWorkspace();
    renderPurchases();
    renderPurchaseCart();

    const currentShopId = state.merchant?.shop_id ?? null;
    if (
      token &&
      state.merchant &&
      currentShopId !== loadedShopId &&
      !loadingPurchasingData
    ) {
      loadPurchasingData();
    }
  };

  document.addEventListener('click', (event) => {
    const tab = event.target.closest('[data-open-tab]');
    if (tab && (tab.dataset.openTab === 'suppliers' || tab.dataset.openTab === 'purchases')) {
      loadPurchasingData();
    }
  });

  loadPurchasingData();
})();